File: Data Structures & Algorithms/queue/submission-1.go
type Node struct {
    val int
    next *Node
    prev *Node
}

func NewNode(val int, prev, next *Node) *Node {
    return &Node { val : val, prev : prev, next : next }
}

type Deque struct {
    head *Node
    tail *Node
}


func NewDeque() *Deque {
    head := NewNode(0,nil,nil)
    tail := NewNode(0,nil,nil)
    tail.prev = head
    head.next = tail
    return &Deque { head:head, tail:tail }
}

func (d *Deque) IsEmpty() bool {
    return d.head.next == d.tail
}

func (d *Deque) Append(value int) {
    newNode := NewNode(value,d.tail.prev,d.tail)
    d.tail.prev.next = newNode
    d.tail.prev = newNode
}

func (d *Deque) AppendLeft(value int) {
    newNode := NewNode(value,d.head,d.head.next)
    d.head.next.prev = newNode
    d.head.next = newNode
}

func (d *Deque) Pop() int {
    var val int = -1
    if !d.IsEmpty() {
        val = d.tail.prev.val
        d.tail.prev.prev.next = d.tail
        d.tail.prev = d.tail.prev.prev
    }
    return val
}

func (d *Deque) PopLeft() int {
    var val int = -1
    if !d.IsEmpty() {
        val = d.head.next.val
        d.head.next.next.prev = d.head
        d.head.next = d.head.next.next
    }
    return val
}

File: Data Structures & Algorithms/lru-cache/submission-1.py
class Node:
    def __init__(self,key :int, val :int) -> None:
        self.key = key
        self.val = val
        self.next = None
        self.prev = None

class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}
        self.head = Node(0,0)
        self.tail = Node(0,0)
        self.head.next = self.tail
        self.tail.prev = self.head

    def insertEnd(self, node : Node) -> None:
        prv,nxt = self.tail.prev, self.tail
        prv.next = nxt.prev = node
        node.next, node.prev = nxt,prv
    
    def remove(self, node : Node) -> None:
        nxt,prv = node.next, node.prev
        nxt.prev, prv.next = prv,nxt

    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            self.remove(node)
            self.insertEnd(node)
            return node.val
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.remove(self.cache[key])

        node = Node(key,value)
        self.cache[key] = node
        self.insertEnd(node)
        
        if len(self.cache) > self.capacity:
            lru = self.head.next
            self.remove(lru)
            self.cache.pop(lru.key)
        
